### DISCUSSION TITLE
UAC2 stereo mic pops and cracklings

### DISCUSSION THREAD

**Discussion by benishor** [hathach/tinyusb#1760] [Q&A]
Hi,

This is more a question about tinyUSB usage caused by my lack of experience with the library rather than a bug.

I added tinyUSB to a pet project involving an ElectroSmith [Daisy Seed](https://www.electro-smith.com/daisy/daisy) board. I modified the existing examples to get a composite device with UAC2 and 2xCDC. The audio device has a 2 channel RX and a 2 channel TX at a rate of 48KHz and a bit depth of 16bit.

The model of my application is that the audio data is generated in a I2S callback and pushed into a circular buffer in order to be used by the USB stack. Then, in `tud_audio_tx_done_pre_load_cb()` I pop the data from the buffer and send it away with `tud_audio_write()`.

While this works, I get occasional pops and cracklings which are annoying to say the least. I verified that they are not caused by lack of bandwidth by lowering the samplerate to 8KHz and getting the same behaviour. 

I have the feeling that it all boils down to my misuse of the library. For example, I am not sure how much data I should be sending in `tud_audio_write()`.  Could anyone provide some assistance or a basic UAC2 mic example which delivers a simple sine with no pops?

Thank you!

**Comment by benishor** [comment]
I attach an image with the visible pops https://imgur.com/a/Y5IVAZo
I am sending a simple sine to the computer.

**Reply by HiFiPhile** [reply]
You can try PR #2282
